diff --git a/lib/usb/usbdevicefactory.cpp b/lib/usb/usbdevicefactory.cpp
index 75210581..631dc0b3 100644
--- a/lib/usb/usbdevicefactory.cpp
+++ b/lib/usb/usbdevicefactory.cpp
@@ -67,6 +67,10 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 		return 0;
 	}
 
+	pParent->GetDevice ()->LogWrite (LogNotice,
+					"Probing device/interface %s",
+					(const char *) *pName);
+
 	CUSBFunction *pResult = 0;
 
 	if (   pName->Compare ("int9-0-0") == 0
@@ -228,6 +232,12 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 	{
 		pResult->GetDevice ()->LogWrite (LogNotice, "Using device/interface %s", (const char *) *pName);
 	}
+	else
+	{
+		pParent->GetDevice ()->LogWrite (LogWarning,
+						"No driver for device/interface %s",
+						(const char *) *pName);
+	}
 
 	delete pName;
 
@@ -236,7 +246,7 @@ CUSBFunction *CUSBDeviceFactory::GetDevice (CUSBFunction *pParent, CString *pNam
 
 CUSBFunction *CUSBDeviceFactory::GetGenericHIDDevice (CUSBFunction *pParent)
 {
-#ifndef EXCLUDE_USB_TOUCHSCREEN
+#if !defined (EXCLUDE_USB_MOUSE) || !defined (EXCLUDE_USB_TOUCHSCREEN)
 	// Must copy parent function here, because we consume the HID report descriptor,
 	// which is requested again later by the HID Use Page specific driver class.
 	CUSBFunction TempFunction (pParent);
@@ -266,9 +276,10 @@ CUSBFunction *CUSBDeviceFactory::GetGenericHIDDevice (CUSBFunction *pParent)
 		return 0;
 	}
 
-	// If we find a Usage Page (Digitizer) item anywhere in the HID report descriptor,
-	// then use the touch screen driver, otherwise the gamepad standard driver.
+	// Generic HID interfaces do not provide a boot protocol from which the
+	// device type can be selected. Inspect the report descriptor instead.
 	const u8 *pDesc = ReportDescriptor;
+	u32 nUsagePage = 0;
 	for (u16 nDescSize = usReportDescriptorLength; nDescSize;)
 	{
 		u8 ucItem = *pDesc++;
@@ -304,11 +315,25 @@ CUSBFunction *CUSBDeviceFactory::GetGenericHIDDevice (CUSBFunction *pParent)
 
 		ucItem &= 0xFC;
 
-		if (   ucItem == 0x04		// Usage Page (Digitizer)
-		    && nArg == 0x0D)
+		if (ucItem == 0x04)		// Usage Page
+		{
+			nUsagePage = nArg;
+		}
+#ifndef EXCLUDE_USB_MOUSE
+		else if (   ucItem == 0x08	// Usage (Mouse)
+			 && nUsagePage == 0x01	// Generic Desktop
+			 && nArg == 0x02)
+		{
+			return new CUSBMouseDevice (pParent);
+		}
+#endif
+#ifndef EXCLUDE_USB_TOUCHSCREEN
+		else if (   ucItem == 0x08	// Usage
+			 && nUsagePage == 0x0D)	// Digitizer
 		{
 			return new CUSBTouchScreenDevice (pParent);
 		}
+#endif
 	}
 #endif
 
